### ISSUE TITLE
Knot doesn't forward all messages, resulting in hanging notes, jittery clock

### ISSUE THREAD

**Issue by narayb** [intechstudio/knot#14]
Reported originally by user named neon on our Discord. Ticket created after reproducing bug in a similar setup. Tested with release FW.

[Video](https://www.youtube.com/shorts/p3hEpaD7vbk)

## Setup used

The setup consists of a MIDI keyboard sending notes into the IN port of Knot and Knot forwarding that into the HOST port. In this case a synth is hosted in the port, Knot is powering it.

Original setup was tested with multiple MIDI keyboards and synths. Reproduced with Atruria Keystep + Audiothingies Micromonsta 2 (see video).

## Steps to reproduce

After creating the setup, pressing more than two keys at the same time and releasing them simultaneously would result in hanging notes 30% of the time.

After testing more, the issue could also be reproduced with just one note + pressure from the Keystep in 80% of tries.

## Conclusions

When multiple messages are received at the IN port, Knot sometimes struggles to process these and misses some in the process. The messages *are* properly sent each time from the Keystep as illustrated by the monome norns on the side, showing us a MIDI monitor on the video footage.

**Comment by narayb** [comment]
Bug confirmed still active on current nightly FW. When doing so, MIDI monitor shows no NOTE OFF messages received.

**Comment by narayb** [comment]
A user added his own experience about lag with Knot on the [forums](https://forum.intech.studio/t/experiencing-some-lag-with-knot/94) and we received user reports of this issue on Support as well, using a MIDI Fighter Twister for sending note on messages.

**Comment by narayb** [comment]
Firmware update v1.0.2 mitigated the issue and one use-case is reportedly stuck-note-free. There still are use-cases that experience the issue however.

- Some USB devices when hosted in the USB HOST port can create stuck notes when controlling a MIDI instrument through the MIDI OUT port e.g. MIDI Fighter 64. Bug not present on Grid.
- Some users experienced jitter/MIDI delay when passing clock through Knot. Investigating.
- A user reported a strange message sent by Knot on a bootup through the MIDI OUT port, hosted device or not. Investigating.

**Comment by jmickelin** [comment]
I have been experiencing the problem in the first bullet point as well. Sadly, updating to the latest firmware (v1.0.4) does not help.

The hardware I have tested it with are the following:
* Controllers (connected via USB to the USB HOST port):
  - My own Korg MicroKey Air (S/N: MICROKEY2-37AIR), confirmed to be using the latest firmware.
  - A different mini MIDI keyboard, which I believe was an Akai LPK25, though I am not 100% sure.
* Instruments (connected via TRS to the MIDI OUT port):
  - A Dreadbox Nymphes synth, confirmed to be using the latest firmware.
  - A computer running a simple MIDI application to visualize the keys being pressed (`vmpk`). Connected using a ESI MidiMate eX USB-MIDI interface between the Knot's MIDI OUT and the computer's USB.

All combinations of setups exhibited the problem with stuck notes as soon as the signal was running through Knot. Connecting the MIDI keyboards directly via USB to the computer worked without issue. Sadly I do not own any non-USB controllers to similarly test the Nymphes directly without the Knot.

---

I have been meaning to file a bug report for this for a long time, but never got around to it. So I'm happy to see that others have done so! Hopefully I can at least provide an additional data point to help figure out the issue.

**Comment by SukuWc** [comment]
Jitter issue fixed, now merged to stable branch

**Comment by jmickelin** [comment]
I'm still getting stuck notes on the latest Nightly firmware (`2025-01-29-1113`), with the same setup as above. It occurs whenever I press and release a handful of keys at the same time. At least one of them is likely to be stuck until I hit the same key again.

**Comment by narayb** [comment]
No, your update has worked perfectly. The patch only addressed the clock jitter issue and has done nothing to the stuck notes. This is an unfortunate case of two issues in one, I will open a new issue for stuck notes exclusively.

**Comment by jmickelin** [comment]
Ah, I see! Thank you :)
